runtime/lua/vim/diagnostic.lua

```diff
@@ -2908,20 +2943,26 @@ local hl_map = {
 function M.status(bufnr)
   vim.validate('bufnr', bufnr, 'number', true)
   bufnr = bufnr or 0
+  local config = assert(M.config()).status or {}
   local counts = M.count(bufnr)
-  local user_signs = vim.tbl_get(M.config() --[[@as vim.diagnostic.Opts]], 'status', 'text') or {}
-  local signs = vim.tbl_extend('keep', user_signs, { 'E', 'W', 'I', 'H' })
-  local result_str = vim
-    .iter(pairs(counts))
-    :map(function(severity, count)
-      return ('%%#%s#%s:%s'):format(hl_map[severity], signs[severity], count)
-    end)
-    :join(' ')
-
+  local format = config.format or default_status_signs
+  vim.validate('diagnostic.status.format', format, { 'table', 'function' })
```

Comment:
I would move this validation under the one for `bufnr` and check that the provided `config.format` setting is either a table or a function.